sycl/source/detail/scheduler/commands.cpp

```diff
@@ -2558,13 +2556,20 @@ getCGKernelInfo(const CGExecKernel &CommandGroup, ContextImplPtr ContextImpl,
     DeviceImageImpl = SyclKernelImpl->getDeviceImage();
     EliminatedArgMask = SyclKernelImpl->getKernelArgMask();
   } else {
-    ur_program_handle_t UrProgram = nullptr;
-    std::tie(UrKernel, std::ignore, EliminatedArgMask, UrProgram) =
+    FastKernelCacheValPtr FastKernelCacheVal =
         sycl::detail::ProgramManager::getInstance().getOrCreateKernel(
             ContextImpl, DeviceImpl, CommandGroup.MKernelName,
             CommandGroup.MKernelNameBasedCachePtr);
+    UrKernel = FastKernelCacheVal->MKernelHandle;
+    EliminatedArgMask = FastKernelCacheVal->MKernelArgMask;
+    // UrProgram/UrKernel are used after KernelCacheVal is destroyed, so caller
+    // must call ur*Release.
+    ContextImpl->getAdapter()->call<UrApiKind::urProgramRetain>(
+        FastKernelCacheVal->MProgramHandle);
+    ContextImpl->getAdapter()->call<UrApiKind::urKernelRetain>(UrKernel);
```

Comment:
Why do we need to Retain now, and did not need to do it prior to your changes?

Reply:
I think, instead of storing `UrKernel` and `FastKernelCacheVal->MProgramHandle` in a separate vectors, we could have a single vector and store `FastKernelCacheVal` (which is a `std::shared_ptr`) in it. In that case we do not need to manually call `urProgramRetain`/`urKernelRetain` APIs